src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -711,6 +711,41 @@ public function release_bookmark( $name ) {
 	}
 
 
+	/**
+	 * Allow internal WordPress code to extract raw HTML snippets associated with a matched tag.
+	 *
+	 * This is meant for internal use to support existing filters that pass segments of input HTML
+	 * for further analysis. Filters which rely on these values should be updated to interact with
+	 * the structural methods this class provides.
+	 *
+	 * @access private
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $internal_use_only Required to accept liability for exposing system internals through this function.
+	 * @param string $content           What portion of the raw HTML markup to return: 'entire-tag' or 'only-attributes'.
+	 * @return string|null Raw HTML markup from input document for matched tag, if matched, else NULL.
+	 */
+	public function _wp_internal_extract_raw_token_markup( $internal_use_only, $contents = 'entire-tag' ) {
+		$disclaimer = <<<INTERNAL_ONLY
+			I understand that this is only for internal WordPress usage and something
+			will likely break if used elsewhere. This function comes with no warranty.
+INTERNAL_ONLY;
+
+		if ( $internal_use_only !== $disclaimer || null === $this->tag_name_starts_at ) {
+			return null;
+		}
```

Comment:
😅

Reply:
I'm not following the meaning of your emoji.

The point I'm trying to make here is that I really want the API to discourage use of this function. It's dangerous because it carries the legacy of interacting with HTML as a string that led to all the issues that brought us here.

In fact the _only_ reason I'm considering this is a purely pragmatic backwards-compatibility issue. Core (and Jetpack and many other popular plugins) expose full or near-full HTML syntax through filters so that extending code can do further processing. For example, a full link HTML is passed into `wp_targeted_link_rel` and some plugins examine the URL to decide what additional `rel` values to add.

We could reconstruct normative HTML for these filters, but that would involve iterating over the attributes to create strings when we have one already. Maybe it's a worthwhile tradeoff, but I think most of those filters are unused and so if we can pass along a substring of an existing string and not touch it, that will have far less impact.

I'd prefer we find ways to update existing code to work structurally instead stringily. We _don't_ need this function today; it's merely an aid to refactors that need the HTML API. If we can deprecate all those filters and leave time to update, then we can avoid this entire thing.

Reply:
Thanks a lot for the added context and explanation, much appreciated!

---

The emoji meant something to the effect, "Wowzers, this one blows [the one we use for the constructor](https://github.com/WordPress/wordpress-develop/blob/6f74f616b8973fdef00e81e0984d33c31b229371/src/wp-includes/html-api/class-wp-html-processor.php#L1400) out of the water. I guess that's the only precedent we have, but it's probably warranted."

Reply:
Well I did realize that I can't make a class const private, so the one in the constructor could be easily foiled by calling `WP_HTML_Processor::CONSTRUCTOR_UNLOCK_CODE`.

In that case I'm less worried anyway because it's there to alert someone why something broke. In this case it's to try and put pressure against dangerous programming habits that _will_ work if attempted, at least to a point.

Reply:
@bor0 there is no code change required by this PR for other plugins and code authors. the point is to preserve backwards compatibility. however, here is one of many examples of how existing filters rely on the raw markup.

```php
// formatting.php
$rel = apply_filters( 'wp_targeted_link_rel', 'noopener', $link_html );
```

here, the `$link_html` is supposed to represent the full `LINK` element's HTML, such as `<link rel="stylesheet" href="some/url">`

lots of existing code then parses this with a regex or string-find, and breaks down in the same cases that normal HTML parsing fails: boolean attribute values, different quoting than expected, duplicate or differently-ordered attributes, escaping issues, etc…

ideally we would move Core towards a future where instead of passing around these strings we could pass around a kind of read-only or otherwise limited HTML Tag Processor, so that filters could reliably read attribute values, but changing the filter would mean breaking those existing plugins